### ISSUE TITLE
ST_PixelAsCentroids

### ISSUE THREAD

**Issue by njmattes** [RDCEP/EDE#20]
@ricardobarroslourenco @legendOfZelda: Is it possible to make `ST_PixelAsCentroids` return *all* bands from a raster? I've been playing with something like the following to return info from the database in a format that can be converted to JSON in the python middleware, so that's useful for drawing the GUI.

```
CREATE OR REPLACE FUNCTION get_raster(mid int, vid int, band int)
RETURNS TABLE(val double precision, st_astext text) AS $$
BEGIN
  RETURN QUERY
  SELECT r.val, ST_AsText(r.geom)
  FROM (SELECT (ST_PixelAsCentroids(rast, band)).*
        FROM grid_data AS gd
        WHERE gd.meta_id = mid
              AND gd.var_id = vid
       ) AS r;
END;
$$ LANGUAGE plpgsql;
```

This seems to work alright, but only gives a single band. It's possible to draw a single band, and then reissue a request (or multiple requests) for the rest of the bands. 

Any thoughts about the best (and also reasonably fast) way to retrieve this? Bearing in mind that, as far as the GUI is concerned, we don't need to return the globe—only the pixels that fill the current viewport.

**Comment by ricardobarroslourenco** [comment]
@njmattes is this table `grid_data` stored at the current database `ede`? I was trying to look at it. 
Perhaps we could iterate over the bands. Or dump, if it is just a column...

**Comment by njmattes** [comment]
@ricardobarroslourenco Looks like the database on the server is a couple steps behind the `develop` branch. I've been testing locally today (was at a kids chess tourney all day). The `netcdf_data` table on the server has a similar layout though.

I think the bands are tucked away inside the binary `rast` column (by tile). Which is what's making things a bit difficult. I think we're trying to stick to the procedures in PostGIS right now, to retrieve the info from the rasters. But `ST_PixelAsCentroids` doesn't actually return *which* band you requested. So you can't return all the bands this way, and then sniff out which ones you want.

Is it necessary to write some lower-level functionality to create a function that returns the rasters as eg `EDE_PixelAsCentroids` with an integer column representing which band each centroid / value pair belongs to?

**Comment by legendOfZelda** [comment]
let's see...

**Comment by njmattes** [comment]
Sorry about that—my sessions should be gone now.

**Comment by ricardobarroslourenco** [comment]
Yeah. Me either (I've just dropped it)

**Comment by legendOfZelda** [comment]
we've basically got these options:

* call ST_PixelAsCentroids in a for loop where in each iteration we use another band number
* we call ST_PixelAsCentroids once to just get the coordinates + in a for loop to get only the values
* we use ST_PixelAsCentroids to get the coordinates + ST_DumpValues to just get the values (for all specified bands which works since ST_DumpValues allows us to specify multiple bands)

more precisely:

* we issue multiple requests 
`SELECT val, ST_AsText(geom) FROM (SELECT (ST_PixelAsCentroids(rast, band_num)).* FROM grid_data where meta_id=1 and var_id=1) centroids;`
in a for loop where `band_num` runs over the band numbers / times we want

* we use
`SELECT x, y, ST_AsText(geom) FROM (SELECT (ST_PixelAsCentroids(rast)).* FROM grid_data where meta_id=1 and var_id=1) centroids;`
to get just the centroids and then
`SELECT x, y, val FROM (SELECT (ST_PixelAsCentroids(rast, band_num)).* FROM grid_data where meta_id=1 and var_id=1) centroids;`
to get the values in a for loop looping over `band_num`

* we use
`SELECT x, y, ST_AsText(geom) FROM (SELECT (ST_PixelAsCentroids(rast)).* FROM grid_data where meta_id=1 and var_id=1) centroids;`
to get just the centroids and then
`SELECT (ST_DumpValues(rast)).* FROM grid_data where meta_id=9 and var_id=2;`
to just get the values

i suggest we use the last one to avoid looping.

**Comment by ricardobarroslourenco** [comment]
@legendOfZelda nice. I was thinking on this last option (without specifying a band) when posted before.

**Comment by legendOfZelda** [comment]
yes, that's what i thought you meant, cool :+1:

**Comment by njmattes** [comment]
Some more questions:

* Do the 10 &times; 10 tiles (always) move left-to-right, top-to-bottom in ? Ie, is `x = 1, y = 1` the top left corner of the data?
* Is the output of `ST_DumpValues` (always) ordered left-to-right, top-to-bottom? Otherwise there's no way to associate a 10 &times; 10 tile of values with a pixel. 
* I'm seeing a lot of `{{ NULL, ... }, .. }` entries. Isn't the default behavior of `ST_DumpValues` not to return `NULL` values? 
* If we're returning `NULL` values from `ST_DumpValues` do we also need to return `NULL` values from `ST_PixelAsCentroids` so the values align?
* If/when we have datasets with irregular resolution, mightn't this strategy begin to fail? Not that we need to care about that presently ...

**Comment by ricardobarroslourenco** [comment]
The 10x10 tiles I think it is something related to an OGC compliance standard, no? When talking about the rotation order, this always comes on mind.

About the raster type, I think this goes into that discussion of data regularization that Raffaelle brought at the all hands meeting. I think this will need to be done prior to ingestion, and looking into some things such as the downscale methods that Joshua uses.

(Sorry by the weird answer, I'm typing at my phone)

**Comment by legendOfZelda** [comment]
under construction...

**Comment by legendOfZelda** [comment]
@njmattes why would we want to all timeframes of a raster inside python? why wouldn't we instead write out the raster with all its timeframes as a GeoTIFF as described here:
http://gis.stackexchange.com/questions/78086/how-to-export-raster-from-postgis-enabled-db-as-gtiff-with-gdal-translate (it works, just tested it)
and then send that as an ordinary file to the user?

**Comment by legendOfZelda** [comment]
this problem is now solved, see query (5) in #23
